tianshou/data/batch.py

```diff
@@ -263,18 +264,37 @@ def __init__(self,
                  **kwargs) -> None:
         if copy:
             batch_dict = deepcopy(batch_dict)
-        if _is_batch_set(batch_dict):
-            self.stack_(batch_dict)
-        elif isinstance(batch_dict, (dict, Batch)):
-            for k, v in batch_dict.items():
-                if isinstance(v, dict) or _is_batch_set(v):
-                    self.__dict__[k] = Batch(v)
-                else:
-                    if isinstance(v, list):
-                        v = np.array(v)
-                        if not issubclass(v.dtype.type, (np.bool_, np.number)):
-                            v = v.astype(np.object)
-                    self.__dict__[k] = v
+        if batch_dict is not None:
+            if isinstance(batch_dict, (dict, Batch)):
+                for k, v in batch_dict.items():
+                    if isinstance(v, (list, tuple, np.ndarray)):
+                        v_ = None
+                        if isinstance(v, np.ndarray):
+                            v_ = v
+                        elif all(isinstance(e, torch.Tensor) for e in v):
+                            v_ = torch.stack(v)
+                            self.__dict__[k] = v_
+                            continue
+                        else:
+                            v_ = np.array(v)
+                        if v_.dtype != np.object:
+                            v = v_  # normal data list, this is the main case
+                            if not issubclass(v.dtype.type,
+                                              (np.bool_, np.number)):
+                                v = v.astype(np.object)
+                        else:
+                            if _is_batch_set(v):
+                                v = Batch(v)  # list of dict / Batch
+                            else:
+                                # this is actually a data list with objects
+                                v = v_
+                        self.__dict__[k] = v
+                    elif isinstance(v, dict):
+                        self.__dict__[k] = Batch(v)
+                    else:
+                        self.__dict__[k] = v
+            elif _is_batch_set(batch_dict):
+                self.stack_(batch_dict)
```

Comment:
This time in test_dqn.py, the calling counts of _is_batch_set reduces from 150k to 3.5k.

Reply:
Yes, nice job! (if it improving the overall performance, but I guess so from your comment)